### PR TITLE
Drop support for Python 3.7

### PR DESCRIPTION
**Pull request by jorenham**
Also removed old python 3.6 code, and increased  the upper bound of some deps

### CODE REVIEW

**Comment by jorenham** [comment]
The CI seems to be complaining about issues irrelevant to this particular PR.

**Comment by jorenham** [comment]
The failing tests can be replicated by checking out the master branch and a `poetry update`.

**Comment by koxudaxi** [comment]
@jorenham 
Thank you for creating the PR.
Before we merge the PR. I want to check the active users of Python 3.7.
I understand the version is EOL. But I guess some projects may use the library in Python 3.7

**Comment by koxudaxi** [comment]
I have run the cli to get the activity of this month.
`$ pypistats python_minor datamodel-code-generator  -t -f markdown`

| category | percent | downloads |
|:---------|--------:|----------:|
| 3.9      |  28.83% |   123,761 |
| null     |  23.56% |   101,128 |
| 3.8      |  19.61% |    84,168 |
| 3.10     |  11.75% |    50,454 |
| 3.7      |  11.39% |    48,904 |
| 3.11     |   4.84% |    20,791 |
| 3.6      |   0.01% |        56 |
| 3.3      |   0.00% |         8 |
| 3.12     |   0.00% |         3 |
| Total    |         |   429,273 |

Date range: 2023-09-01 - 2023-09-24

**Comment by jorenham** [comment]
@koxudaxi It's rather sad to see that there are more that twice the amount of 3.7 users than 3.11...

**Comment by koxudaxi** [comment]
@jorenham 

> @koxudaxi It's rather sad to see that there are more that twice the amount of 3.7 users than 3.11...

This is the real world😅 

We should wait for `pydantic` itself to drop 3.7.
This is probably because `pydantic` is often used in the same project.
https://github.com/pydantic/pydantic/pull/7188

**Comment by jorenham** [comment]
@koxudaxi That makes sense. I'll try again in a couple of years then 😛

**Comment by jorenham** [comment]
@koxudaxi https://github.com/pydantic/pydantic/pull/7188 has been merged; time flies!

File: datamodel_code_generator/format.py [hunk 1]
```diff
@@ -12,8 +12,6 @@
 
 
 class PythonVersion(Enum):
-    PY_36 = '3.6'
-    PY_37 = '3.7'
```
Comment (koxudaxi) [inline datamodel_code_generator/format.py:22]: @jorenham 
I agree with you it's to drop support 3.7 from runtime support list.
Should we delete PY_37 for output?
Personally, I'm happy to drop the old version from the output model list.
But I want to think about it again. :thinking:

**Comment by koxudaxi** [comment]
$ pypistats python_minor datamodel-code-generator  -t -f markdown
| category | percent | downloads |
|:---------|--------:|----------:|
| 3.9      |  28.87% |    95,632 |
| null     |  21.40% |    70,884 |
| 3.8      |  18.20% |    60,308 |
| 3.10     |  13.32% |    44,143 |
| 3.7      |   9.83% |    32,564 |
| 3.11     |   7.83% |    25,935 |
| 3.12     |   0.53% |     1,750 |
| 3.6      |   0.02% |        67 |
| 3.3      |   0.01% |        20 |
| Total    |         |   331,303 |

Date range: 2023-11-01 - 2023-11-18

**Comment by koxudaxi** [comment]
`0.24.*`
image

**Comment by koxudaxi** [comment]
`0.23.*`
image

**Comment by koxudaxi** [comment]
```bash
pypistats python_minor datamodel-code-generator -t -f markdown

| category | percent | downloads |
| :--------|-------: |---------: |
| 3.10     |  34.49% |   259,986 |
| 3.11     |  21.15% |   159,453 |
| null     |  18.45% |   139,114 |
| 3.9      |  11.02% |    83,073 |
| 3.8      |   7.09% |    53,462 |
| 3.12     |   6.52% |    49,118 |
| 3.7      |   1.27% |     9,543 |
| 3.6      |   0.00% |        37 |
| 3.13     |   0.00% |        25 |
| Total    |         |   753,811 |

Date range: 2024-06-01 - 2024-06-27
```

**Comment by koxudaxi** [comment]
@jorenham 
I want to drop support Python 3.7
but the PR needs to be updated. 
I'm sorry. I will close the PR and recreate a new PR for it.

**Pull request closed without merging**